infrastructure/aws/workflows/deploy.yml

```diff
@@ -38,23 +38,31 @@ jobs:
           } >> $GITHUB_OUTPUT
         id: extract_branch
 
+      - name: Extract environment name
+        env:
+          ENVIRONMENT: ${{ steps.extract_branch.outputs.branch == 'main' && 'PRODUCTION' || 'STAGING' }}
+        run: echo "environment=$ENVIRONMENT" >> $GITHUB_OUTPUT
+        id: extract_environment
+
       - name: Copy env variables to docker
         if: ${{ github.event_name == 'workflow_dispatch' || steps.client-changes.outputs.client == 'true' }}
         run: |
-          echo "${{ steps.extract_branch.outputs.branch == 'main' && secrets.PRODUCTION_CLIENT_ENV_FILE || secrets[format('{0}_CLIENT_ENV_FILE', steps.extract_branch.outputs.branch_upper)] }}" > client/.env.local
+          # Filter client secrets & vars and save .env file (names starting with (TF_)(PRODUCTION|STAGING_)CLIENTENV_
+          echo '${{ secrets != null && toJSON(secrets) || null }}' '${{ vars != null && toJSON(vars) || null}}' | jq -r 'to_entries | map(.key + "=" + .value) | .[]' | grep -E "^(TF_)?(${{ steps.extract_environment.outputs.environment }}_)?CLIENTENV_" | sed -r 's/^(TF_)?(${{ steps.extract_environment.outputs.environment }}_)?CLIENTENV_//' > client/.env.local
```

Comment:
the steps are piped into a single command to avoid writing anything to intermediate temp files; agree each step should be descibed in terms of input / output :+1:

Reply:
I was thinking writing the output of each step to an env var. I don't think a "1 step = 1 command = 1 env var" level of granularity is needed, but something in between baby steps and "endless train of debug hell" would be nice - along with docs, yes